### PR TITLE
Static deepcopy – phase 1

### PR DESCRIPTION
**Pull request by sttts**
This PR is the follow-up of https://github.com/kubernetes/kubernetes/pull/36412, replacing the
dynamic reflection based deepcopy with static DeepCopy+DeepCopyInto methods on API types.

This PR **does not yet** include the code dropping the cloner from the scheme and all the
porting of the calls to scheme.Copy. This will be part of a follow-up "Phase 2" PR.

A couple of the commits will go in first:
- [x] audit: fix deepcopy registration  https://github.com/kubernetes/kubernetes/pull/48599
- [x] apimachinery+apiserver: separate test types in their own packages #48601 
- [x] client-go: remove TPR example #48604
- [x] apimachinery: remove unneeded GetObjectKind() impls #48608 
- [x] sanity check against origin, that OpenShift's types are fine for static deepcopy https://github.com/deads2k/origin/pull/34

TODO **after** review here:
- [x] merge https://github.com/kubernetes/gengo/pull/32 and update vendoring commit

### CODE REVIEW

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go [hunk 1]
```diff
@@ -136,20 +135,6 @@ func (b *blockingBackend) processEvents(ev ...*auditinternal.Event) error {
 	return b.w.RestClient.Post().Body(&list).Do().Error()
 }
 
-// Copied from generated code in k8s.io/apiserver/pkg/apis/audit.
-//
-// TODO(ericchiang): Have the generated code expose these methods like metav1.GetGeneratedDeepCopyFuncs().
-var auditDeepCopyFuncs = []conversion.GeneratedDeepCopyFunc{
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go]: @ericchiang ptal

**Comment by sttts** [comment]
@smarterclayton @deads2k all prerequisites merged. I have split the PR here into a number of independent steps, most of them are boring and pure repititions. Will squash after review of course. ptal.

File: cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.go [hunk 2]
```diff
@@ -32,6 +33,8 @@ type TestType struct {
 	Status TestTypeStatus `json:"status,omitempty"`
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```
Comment (deads2k) [inline cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.go]: it seems weird that you have to specify this interface.  What else would you generate a deep copy for?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/deepcopy.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+func (in *TableRow) DeepCopy() *TableRow {
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/deepcopy.go]: there's a fuzzer that tests this, right?

File: hack/make-rules/helpers/cache_go_dirs.sh [hunk 4]
```diff
@@ -39,7 +39,7 @@ function kfind() {
     # include the "special" vendor directories which are actually part
     # of the Kubernetes source tree - generators will use these for
     # including certain core API concepts.
-    find -H . ./vendor/k8s.io/apimachinery ./vendor/k8s.io/apiserver ./vendor/k8s.io/kube-aggregator ./vendor/k8s.io/apiextensions-apiserver ./vendor/k8s.io/metrics ./vendor/k8s.io/sample-apiserver ./vendor/k8s.io/api \
+    find -H . ./vendor/k8s.io/apimachinery ./vendor/k8s.io/apiserver ./vendor/k8s.io/kube-aggregator ./vendor/k8s.io/apiextensions-apiserver ./vendor/k8s.io/metrics ./vendor/k8s.io/sample-apiserver ./vendor/k8s.io/api ./vendor/k8s.io/client-go \
```
Comment (deads2k) [inline hack/make-rules/helpers/cache_go_dirs.sh]: oh goody.  Won't this make it pick up for the other zz_generated generators too?

File: cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.go (hunk 2 above)
Comment (sttts) [inline cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.go]: Every (api) type gets a DeepCopy func, but we have runtime.Objects and selectors which are used in a polymorphic way. I.e. you get a runtime.Object without knowing which type it really is. For top-level API objects we create an additional type-independent `DeepCopyObject() runtime.Object` method. My original proposal was adding these automatically for those types embedding `TypeMeta` (compare https://github.com/kubernetes/gengo/pull/44). It was decided by @thockin and @smarterclayton that this was too much magic. So we have these annotations now which trigger the creation of `DeepCopyObject() runtime.Object` explicitly.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/deepcopy.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/deepcopy.go]: will double check. They should be part of our generic roundtrip test.

File: staging/src/k8s.io/apimachinery/pkg/watch/mux.go [hunk 5]
```diff
@@ -84,6 +84,12 @@ type functionFakeRuntimeObject func()
 func (obj functionFakeRuntimeObject) GetObjectKind() schema.ObjectKind {
 	return schema.EmptyObjectKind
 }
+func (obj functionFakeRuntimeObject) DeepCopyObject() runtime.Object {
+	if obj == nil {
+		return nil
+	}
+	return obj
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/watch/mux.go]: Comment why this one doesn't actually do a copy.

File: hack/make-rules/helpers/cache_go_dirs.sh (hunk 4 above)
Comment (sttts) [inline hack/make-rules/helpers/cache_go_dirs.sh]: Yes. But none of those produces any file.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/watch.go [hunk 6]
```diff
@@ -80,3 +80,10 @@ type InternalEvent watch.Event
 
 func (e *InternalEvent) GetObjectKind() schema.ObjectKind { return schema.EmptyObjectKind }
 func (e *WatchEvent) GetObjectKind() schema.ObjectKind    { return schema.EmptyObjectKind }
+func (e *InternalEvent) DeepCopyObject() runtime.Object {
+	if c := e.DeepCopy(); c != nil {
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/watch.go]: if else seems weird.  why can't you just return `e.DeepCopy`?

File: staging/src/k8s.io/apimachinery/pkg/watch/mux.go (hunk 5 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/watch/mux.go]: done

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go [hunk 7]
```diff
@@ -145,6 +145,49 @@ func (u *Unstructured) UnmarshalJSON(b []byte) error {
...
+		}
+		return clone
+	case []interface{}:
+		clone := make([]interface{}, len(x))
+		for i := range x {
+			clone[i] = deepCopyJSON(x[i])
+		}
+		return clone
+	default:
+		return x
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: if my map value was `*int64`, how does this make a copy of that?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/watch.go (hunk 6 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/watch.go]: nil interface vs. non-nil interface will nil pointer.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go (hunk 7 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: It's json, it cannot be *int64, only int64, string, map, slice. Adding a comment.

Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: done

**Comment by sttts** [comment]
@liggitt I have run static deepcopy against origin (deads2k/origin#34) and manually verified the types to be supported. The generated code looks sane and mostly even compiles (without any deeper changes). Especially, origin has no additional interfaces we have to extend with `DeepCopyInterfaceName()` funcs, i.e. only those of kube: `runtime.Object` and selectors. Moreover, we only have one custom deepcopy (`ResourceQuotasStatusByNamespace`) which is easily supported with the new deepcopy. So, as far as I can judge now – before a real 1.8 rebase – this looks good.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/deepcopy.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/deepcopy.go]: I have add meta apigroup fuzzing and roundtrip testing here: https://github.com/kubernetes/kubernetes/pull/48810

**Comment by sttts** [comment]
@liggitt @deads2k addressed all comments. anything else before this can go in?

File: cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.go (hunk 2 above)
Comment (deads2k) [inline cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.go]: > It was decided by @thockin  and @smarterclayton that this was too much magic. So we have these annotations now which trigger the creation of DeepCopyObject() runtime.Object explicitly.

Weird.  Seems like a switch that we'll only throw one way.

**Comment by sttts** [comment]
@smarterclayton want to go forward with this. Is it approved from your side? If yes, please merge https://github.com/kubernetes/gengo/pull/32 and I will revendor here.

**Comment by sttts** [comment]
Rebased and revendored.

File: staging/src/k8s.io/api/apps/v1beta1/types.go [hunk 8]
```diff
@@ -58,6 +58,7 @@ type ScaleStatus struct {
 
 // +genclient=true
 // +noMethods=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```
Comment (janetkuo) [inline staging/src/k8s.io/api/apps/v1beta1/types.go]: I think this change broke my PR #48746 which adds a new API version apps/v1beta2:

```
# k8s.io/kubernetes/vendor/k8s.io/api/apps/v1beta2
vendor/k8s.io/api/apps/v1beta2/register.go:47: cannot use Deployment literal (type *Deployment) as type runtime.Object in argument to scheme.AddKnownTypes:
        *Deployment does not implement runtime.Object (missing DeepCopyObject method)
...
```

I guess I can fix this error by adding this line to apps/v1beta2/types.go?

```go
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```

Is it documented anywhere?

Comment (sttts) [inline staging/src/k8s.io/api/apps/v1beta1/types.go]: Yes, can be fixed via that tag. It's document in gengo https://github.com/kubernetes/gengo/blob/master/examples/deepcopy-gen/main.go#L17. Do we have any better place for this kind of docs?

Comment (janetkuo) [inline staging/src/k8s.io/api/apps/v1beta1/types.go]: @caesarxuchao has an open PR documenting required steps when making API changes https://github.com/kubernetes/community/pull/788

**Comment by hongchaodeng** [comment]
Hi @sttts ,

After this change, all APIs, including custom ones, needs to autogen DeepCopy() method. This is not simple. So what's the plan for this? Will it require all APIs to do additional steps to autogen? Could it be embedded into ObjectMeta/ListMeta ?

Thanks in advance for your answer!

**Comment by sttts** [comment]
@hongchaodeng take a look at https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kube-gen. This will soon be exported into its own repository. Moreover, work is being done to make the usage of kube-gen much easier for your own types.

For the time being, an option would be to turn https://github.com/kubernetes/kubernetes/pull/49575/commits/741c740c45224a2d07287b136d753eb3b72fd48c#diff-2f2666f5ac9b57971d6200baba3f2dd0L171 into a **standalone** reflection based deepcopier. PRs are welcome to create such a thing out of the existing code. I can see some value in this for quick prototyping. But this should be strictly independent from apimachinery's runtime, more like a Go util, probably even part of k8s.io/util.

**Comment by lavalamp** [comment]
@sttts @smarterclayton @deads2k I hear this added the requirement to have a DeepCopy method to be a runtime.Object? I would like to have a discussion at least on the sig email list before we change central interfaces like that, I did not know that this was part of this PR and would have probably asked for it not to happen if I had.

**Comment by sttts** [comment]
@lavalamp this part was discussed up and down in the orginal PR https://github.com/kubernetes/kubernetes/pull/36412 and is even explicitly stated in line 2 of its description. Then again this was discussed in the Gengo PR https://github.com/kubernetes/gengo/pull/32. Both PRs asked for reviews for months, were reviewed and eventually merged. Moreover, we had the static deepcopy topic twice in the sig meeting. Not sure how this change should have been much more visible.

**Comment by lavalamp** [comment]
@sttts Yeah I am really sorry, I know you had a really hard time getting attention on that change. I cannot stay on top of everything, and I kept pushing that to the bottom of my list because it was huge. :( I guess I am saying, it would be good to call out top-level interface changes separately.

**Comment by sttts** [comment]
@lavalamp at least I had that very change here in this PR as a separate commit (until squashing before merge) for that reason. Let me know what your concerns are with the DeepCopyObject func in the interface other than cosmetical ones that we add another func to a small interface. This topic was discussed a couple of times off- and online and eventually the conclusion always was that not having that leads us back to the uncertainty of not knowing whether an object can successfully be copied or not. Take a look at https://github.com/kubernetes/kubernetes/pull/48765 to see those many many places where we try to handle the error case. The comments above some of them give quite some confidence that it's a small price to pay to have that additional func in the interface and get the static typing through the go compiler.

The use-case I agree with is that during prototyping the need for the deepcopy-gen is less comfortable than our reflection based fallback cloner. If this really is an issue, we can easily readd a standalone cloner which can explicitly be used if needed: `func (obj) DeepCopyObject() runtime.Object { return reflectioncloner.DeepCopy(obj) }`. My personal opinion though is that kube-gen needs some love to become much easier to apply to custom types. More and more people use CRDs and UASs with their own types, clients and informers.

**Comment by caesarxuchao** [comment]
@sttts I still don't get why we need to add Deepcopy to the `runtime.Object` interface. I'm sorry if this has been mentioned elsewhere. Go compiler will generate errors when DeepCopy() is called, developer can then invoke kube-gen.

**Comment by sttts** [comment]
@caesarxuchao two situations:
1. we have runtime.Objects as fields in a number of API types. Imagine you have `type X struct { obj runtime.Object }`. Now, deepcopy `X`.
1. inside some generic algorithm on runtime.Objects (e.g. in the generic registry, for example the `UpdatedObjectInfo` implementation) you need a deepcopy. How will you do that?

All this with the requirement that you do not want handle the error case that deepcopy fails: deepcopy should work, unconditionally, always, without a chance to fail. For this you have to enforce that a deep copy is possible on each object. The way in Golang to do this is via an interface. In 99% of places of our code this interface is `runtime.Object`.

In theory, we could remove `DeepCopyObject()` from the runtime.Object interface and just define a `DeepCopyAbleObject` interface next to it and then `obj.(DeepCopyAbleObject).DeepCopyObject()`. Remember: for generic code this `DeepCopyObject()` signature must also be uniform. I.e. type casting will not allow you to call `func (*Pod) DeepCopy() *Pod`. So you need this `DeepCopyObject() runtime.Object` shim. And for type casting there isn't a way around an interface either. Eventually you end up with something like DeepCopyObject() in an interface implemented by every runtime.Object. Golang doesn't give us any other option under the do-not-error requirement. The other option of using reflection (= find the `func (*Pod) DeepCopy() *Pod`, call it, type-cast back to runtime.Object) as done by our old cloner can fail – leading to broken, complex error handling – because the go compiler is of no help to enforce the existence of a `DeepCopy()`. Moreover, it is slow.

**Comment by caesarxuchao** [comment]
Thanks for the explanation @sttts. I went to read the original thread. I think the "do not error" + "no fallback with reflection" + "deepcopy runtime.Object" forced us to add the `DeepCopyObject()` to the interface. These requirements all make sense in a server environment.

But in the client environment these requirements don't hold. As you mentioned, clients can use a reflection based tool to do deepcopies. I don't have a concrete idea on how to spare our client users from running kube-gen. The clientsets only rely on the [serializers](https://github.com/kubernetes/apimachinery/tree/master/pkg/runtime/serializer), can we pass around the original `runtime.Object` interface there?

**Comment by liggitt** [comment]
> But in the client environment these requirements don't hold.

I disagree. Clients are just as vulnerable to objects that reflection cannot correctly deep copy. Defaulting to reflection for API objects whose author has not provided a deep copy impl puts us right back where we were previously, with a deep copy impl that literally falls back to shallow copy in some cases.

**Pull request merged**